tls/extensions/s2n_client_psk.c

```diff
@@ -181,19 +170,23 @@ static S2N_RESULT s2n_client_psk_recv_identity_list(struct s2n_connection *conn,
         uint32_t obfuscated_ticket_age = 0;
         GUARD_AS_RESULT(s2n_stuffer_read_uint32(wire_identities_in, &obfuscated_ticket_age));
 
-        /* TODO: Implement the callback to choose a PSK: https://github.com/awslabs/s2n/issues/2397
-         *
-         * When we don't have a callback configured to choose a PSK, we should fall back to accepting
-         * the first PSK identity that also exists in our list of supported PSKs. */
-        GUARD_RESULT(s2n_match_psk_identity(&conn->psk_params.psk_list, &identity, &conn->psk_params.chosen_psk));
-
-        if (conn->psk_params.chosen_psk) {
-            conn->psk_params.chosen_psk_wire_index = wire_index;
-            return S2N_RESULT_OK;
-        }
-
         wire_index++;
     }
+
+    /* Make sure we have a psk_selection_cb before proceeding */
+    if (!conn->config->psk_selection_cb) {
+        GUARD_AS_RESULT(s2n_config_set_psk_selection_callback(conn, s2n_select_psk_identity));
+        ENSURE_REF(conn->config->psk_selection_cb);
+    }
+
+    GUARD_AS_RESULT(conn->config->psk_selection_cb(conn, wire_identities, identities_count,
+                                         &conn->psk_params.chosen_psk_wire_index));
+
```

Comment:
Do we need a safety check to make sure that the callback function the customer sets returns a wire index < identities_count?